Implement `join/leave_multicast_group()` for IPv6

lucasvr opened the pull request:
This patch rebases @jgallagher's PR #602 to the tip of the main branch, adding support for IPv6 multicast groups.

As in the original PR, it is possible to join groups by sending an initial MLDv2 Report packet. In addition to PR #602, this PR also implements `leave_multicast_group()`.

It is not yet possible to resend the change report or respond to queries.

Commit:
Update PR #602: implement `join_multicast_group()` for IPv6

This patch rebases @jgallagher's to the tip of the main branch,
adding support for IPv6 multicast groups. As in the original PR,
it is only possible to join groups by sending an initial MLDv2
Report packet. It is not yet possible to resend the change report,
leave groups, respond to queries, etc.

Commit:
IPv6: implement `leave_multicast_group()`

Also, modify `test_join_ipv6_multicast_group()` so that it joins
the mDNS multicast group instead of 0xfe80::1. Because the lower 24
bits of 0xfe80::01 match the lower part of the IPv6 address assigned
to the test device by `crate::tests::setup()`, `has_multicast_group()`
would always returns `true`.

Commit:
Apply rustfmt on igmp.rs

thvdveld commented on src/iface/interface/igmp.rs:
```diff
@@ -10,6 +10,8 @@ pub enum MulticastError {
     GroupTableFull,
     /// IPv6 multicast is not yet supported.
     Ipv6NotSupported,
```
I think you can remove the `Ipv6NotSupported` now?

lucasvr replied:
Done. Thanks for looking into this!

thvdveld reviewed (changes requested):
I think everything looks good. Just a small thing that I noticed.

lucasvr committed:
Remove `Ipv6NotSupported` from `enum MulticastError`

No part of the code uses this anymore.

thvdveld commented on src/wire/ipv6hbh.rs:
```diff
@@ -105,6 +106,18 @@ impl<'a> Repr<'a> {
     }
+
+    /// The hop-by-hop header containing a PadN and a MLDv2 router alert option.
+    pub fn mldv2_router_alert(n: u8) -> Self {
```
Is there a case where the padding is not equal to PadN(0)?

lucasvr replied:
If the only option being transmitted is `RouterAlert` then `PadN(0)` would suffice. But if one wishes to append more options after `RouterAlert` then a different padding value might be needed.

thvdveld replied:
Shouldn't the padding be calculated at the end, when emitting the options?

lucasvr replied:
To keep the API surface small, I opted to let `mldv2_router_alert()` handle both options. But I agree that going forward it may be good to have a dedicated function to emit the padding option. I have pushed a commit addressing that.

thvdveld commented on Cargo.toml:
```diff
@@ -296,6 +296,10 @@ required-features = ["log", "medium-ethernet", "proto-ipv4", "socket-tcp"]
 
+[[example]]
+name = "multicast6"
+required-features = ["std", "medium-ethernet", "medium-ip", "phy-tuntap_interface", "proto-ipv6", "socket-udp"]
```
Shouldn't `proto-igmp` be required as well?

lucasvr replied:
Since `proto-igmp` belongs to the `default` dependencies I didn't spot this before. Fixed.

thvdveld commented on src/iface/interface/ipv6.rs:
```diff
@@ -472,4 +473,67 @@ impl InterfaceInner {
+                }
+            }
+        })
+    }
```
Could you move this function under the `fn ipv6_addr` function?

lucasvr replied:
Done.

thvdveld commented on src/iface/interface/tests/ipv6.rs:
```diff
@@ -1169,3 +1169,126 @@ fn get_source_address_empty_interface() {
+                },
+            )
+            .unwrap();
+        assert_eq!(hbh_repr, expected_hbh_repr);
```
I'm not sure why this is needed? Are we unable to parse the Hop-by-Hop option?

lucasvr replied:
This was mostly out of habit. I have updated the test to check the output of the parsed representation instead.

thvdveld commented on src/wire/mld.rs:
```diff
@@ -294,6 +294,63 @@ impl<T: AsRef<[u8]> + AsMut<[u8]>> AddressRecord<T> {
+        if self.payload.len() == record_payload.len() {
+            // TODO: handle the case where the payload sizes are different
+            record_payload.copy_from_slice(self.payload);
+        }
```
Should we maybe set the payload outside of this function, since the `buffer_len` function returns the buffer length excluding the payload length?

lucasvr replied:
Good idea -- I have now moved the payload copy to `wire::Repr::emit()`, which fits nicely with the way that `Repr::Query` and `Repr::Report` payloads are handled.

lucasvr committed:
Address code review suggestions

- Handle payload copy outside of mld::AddressRecordRepr::emit()
- On unit test, use the parsed Hop-by-Hop option instead of checking
  that the wire representation of the two structures is the same
- Minor code changes

thvdveld commented on src/wire/mld.rs:
```diff
@@ -386,6 +440,21 @@ impl<'a> Repr<'a> {
+                    }
+                    record.emit(&mut AddressRecord::new_unchecked(&mut *payload));
+                    payload = &mut payload[record.buffer_len()..];
+                }
```
Is this correct? I am not sure what you are trying to achieve with the if block, because I guess that the emit function of the record is going to override some parts of the payload variable.

lucasvr replied:
This was actually part of @jgallagher's original PR and, indeed, it can go. The source payload is always empty in the current implementation and, even if it weren't, the copied data would be overwritten when emitting the address record. I have just fixed this. Thanks for the code review!

lucasvr committed:
Address code review

- Split mldv2_router_alert() in two, adding a new function named
  push_padn_option(u8) to emit a PadN hop-by-hop option after any
  other options have been pushed already
- When emiting ReportRecordReprs, remove the conditional block
  that attempted to copy payload data over the destination record

lucasvr committed:
Update unit test

lucasvr committed:
rustfmt
